### PR TITLE
Add Classification Code to new disabilities and validate names

### PR DESCRIPTION
**Pull request by markgreenburg**
## Description
- Adds `classificationCode` to `newDisabilities` in the submit transformer
- Adds FE validation on new disability names, ensuring that names which are not on the pre-approved list are tested against VDC regex
- Adds tests for the new submit transformer functionality as well as the FE input validator

## Testing done
- Tested locally
- Added unit tests

## Screenshots
Does not pass regex and not on the disability list:
screen shot 2019-01-22 at 3 39 05 pm

-----

Does not pass validation but *is* on the disability list:
screen shot 2019-01-22 at 3 39 19 pm

## Acceptance criteria
- [x] Add classification code where available to submitted new disabilities
- [x] Add FE conditional validation for new disability names
- [x] Test functionality

## Definition of done
- [x] Events are logged appropriately
- [x] Documentation has been updated, if applicable
- [x] A link has been provided to the originating GitHub issue (or connected to it via ZenHub)
- [x] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

**Comment by markgreenburg** [comment]
@peggygannon need your help on content for the error message. Technically right single quotes (’) are allowed but not apostrophes ('), I omitted both from the 'allowed' message to avoid confusion.

**Comment by markgreenburg** [comment]
@djmunoz1 @goldenmeanie EVSS let us know that VDC accepts right single quotes (’) but not apostrophes ('). Our choices are to either let people know that apostrophes aren't allowed (which honestly I'd prefer from a technical perspective) or transform the user's input to make apostrophes into single quotes in the pre-submit step. I'd like to briefly discuss  pros & cons of each approach in standup holdover tomorrow.

File: src/applications/disability-benefits/all-claims/validations.js [hunk 1]
```diff
@@ -205,3 +210,14 @@ export const hasMonthYear = (err, fieldData) => {
...
   }
 };
+
+export const validateDisabilityName = (err, fieldData) => {
+  if (
+    !LOWERED_DISABILITY_DESCRIPTIONS.includes(fieldData.toLowerCase()) &&
+    !VDC_DISABILITY_NAME_REGEX.test(fieldData)
+  ) {
+    // technically single quotes (’) are allowed as well but leaving out of
+    // this message to avoid confusing veterans who can't tell the difference
```
Comment (markgreenburg) [inline src/applications/disability-benefits/all-claims/validations.js]: TBD what we do with this, but if we end up wanting to allow apostrophes and transform them to single quotes in the submit transformer, I'd rather submit this as-is and then push up another PR for that functionality separately.

**Comment by goldenmeanie** [comment]
We cannot put this on the Veteran. How heavy is transform `unicode-foo` to `unicode-bar`?

**Comment by markgreenburg** [comment]
@goldenmeanie the transformation itself isn't that hard to do right before we submit, the problem comes up when we try to use the disability name to match against something else in a different pre-submit step, like the list of known disabilities that uses only regular apostrophes. It's possible, but definitely needs to be thought through a bit more on our end. Is it acceptable to disallow apostrophes in the meantime?

**Comment by markgreenburg** [comment]
@peggygannon no caps necessary. They can use any letters or numbers, but can only use the following special characters: , . / \ ( )
(and also a single quote (’), not to be confused with an apostrophe, but we're still working out the details of how to best implement this without putting the burden on the veteran).

**Comment by goldenmeanie** [comment]
@markgreenburg could the compare ignore punctuation entirely? Also, you were right, this does seem like a slack or live conversation :)

**Comment by cvalarida** [comment]
Ooh, I like this idea. @markgreenburg, we could make a `normalizeConditionName` function that does whatever we want it to and keep consistency.
```js
// Use the normalized name as the object property in claimedConditions
claimedConditions[normalizeConditionName(disability.condition)] = true;

// Check that each entry in the claimedConditions matches a disability name in one of the lists
Object.keys(claimedConditions).forEach(name => {
  const matchesRatedCondition = ratedDisabilities
    .some(disability => normalizeConditionName(disability.name) === name);
};
```
I was originally thinking we'd then be able to safely make the transformation before sending it to the API, but it seems like it should be the API's responsibility to massage this data so it fits the third-party dependency's (EVSS') validation.

We'll need to figure out where all we compare condition names, though. I think maybe just for `treatedDisabilityNames`.

**Comment by peggygannon** [comment]
@markgreenburg Suggested error message below. I kept it simple, so as to not confuse the Veteran. 
Let me know if you think this message would work: 

Please type in your condition again. The system can't accept apostrophe marks.

File: src/applications/disability-benefits/all-claims/submit-transformer.js [hunk 2]
```diff
@@ -241,6 +243,43 @@ export function transform(formConfig, form) {
...
+
+    const flippedDisabilityLabels = {};
+    Object.entries(disabilityLabels).forEach(([code, description]) => {
+      flippedDisabilityLabels[description.toLowerCase()] = code;
+    });
+
+    const newDisabilitiesWithClassificationCodes = newDisabilities.map(
+      disability => {
+        const { condition } = disability;
+        if (!condition) {
```
Comment (cvalarida) [inline src/applications/disability-benefits/all-claims/submit-transformer.js]: Why would a disability not have a `condition`?

File: src/applications/disability-benefits/all-claims/submit-transformer.js [hunk 3]
```diff
@@ -241,6 +243,43 @@ export function transform(formConfig, form) {
...
+          return disability;
+        }
+        const loweredDisabilityName = condition.toLowerCase();
```
Comment (cvalarida) [inline src/applications/disability-benefits/all-claims/submit-transformer.js]: What do you think about standardizing on a `normalizeDisabilityName` or something? Even if for now that is just
```js
const normalizeDisabilityName = name => name.toLowerCase();
```

File: src/applications/disability-benefits/all-claims/validations.js (hunk 1 above)
Comment (cvalarida) [reply src/applications/disability-benefits/all-claims/validations.js]: Yeah, this sounds fine to me.

File: src/applications/disability-benefits/all-claims/validations.js [hunk 4]
```diff
@@ -205,3 +210,14 @@ export const hasMonthYear = (err, fieldData) => {
...
+    err.addError('The only special characters allowed are: , . ( ) / \\');
```
Comment (cvalarida) [inline src/applications/disability-benefits/all-claims/validations.js]: Breaking news! :newspaper: 

The `\` isn't allowed and the `-` is.

**Review by cvalarida — APPROVED** [review]
Looks like the main thing is using the new regex, but it otherwise looks good.

**Pull request merged**